rust/sdk-processor/src/utils/starting_version.rs

```diff
@@ -1,52 +1,278 @@
+/// or `staring_version` from the config, or 0 if not set.
+///
+/// If this is a backfill processor and threre is an in-progress backfill, this will return
+/// the checkpointed version + 1.
+///
+/// If this is a backfill processor and there is not an in-progress backfill (i.e., no checkpoint or
+/// backfill status is COMPLETE), this will return `starting_version` from the config, or 0 if not set.
 pub async fn get_starting_version(
     indexer_processor_config: &IndexerProcessorConfig,
     conn_pool: ArcDbPool,
 ) -> Result<u64> {
-    // If starting_version is set in TransactionStreamConfig, use that
-    if indexer_processor_config
-        .transaction_stream_config
-        .starting_version
-        .is_some()
-    {
-        return Ok(indexer_processor_config
-            .transaction_stream_config
-            .starting_version
-            .unwrap());
-    }
-
-    // If it's not set, check if the DB has latest_processed_version set and use that
-    let latest_processed_version_from_db =
+    // Check if there's a checkpoint in the approrpiate processor status table.
+    let latest_processed_version =
         get_latest_processed_version_from_db(indexer_processor_config, conn_pool)
             .await
             .context("Failed to get latest processed version from DB")?;
-    if let Some(latest_processed_version_tracker) = latest_processed_version_from_db {
-        // Add +1 to start from the version after the last successful version
-        return Ok(latest_processed_version_tracker + 1);
-    }
 
-    // If latest_processed_version is not stored in DB, return the default 0
-    Ok(0)
+    // If nothing checkpointed, return the `starting_version` from the config, or 0 if not set.
+    Ok(latest_processed_version.unwrap_or(
+        indexer_processor_config
+            .transaction_stream_config
+            .starting_version
+            .unwrap_or(0),
+    ))
 }
 
-/// Gets the start version for the processor. If not found, start from 0.
-pub async fn get_latest_processed_version_from_db(
+async fn get_latest_processed_version_from_db(
     indexer_processor_config: &IndexerProcessorConfig,
     conn_pool: ArcDbPool,
 ) -> Result<Option<u64>> {
     let mut conn = conn_pool.get().await?;
 
-    match ProcessorStatusQuery::get_by_processor(
+    if let Some(backfill_config) = &indexer_processor_config.backfill_config {
+        let backfill_status = BackfillProcessorStatusQuery::get_by_processor(
+            &backfill_config.backfill_alias,
+            &mut conn,
+        )
+        .await
+        .context("Failed to query backfill_processor_status table.")?;
+
+        // Return None if there is no checkpoint or if the backfill is old (complete).
+        // Otherwise, return the checkpointed version + 1.
+        return Ok(
+            backfill_status.and_then(|status| match status.backfill_status {
+                BackfillStatus::InProgress => Some(status.last_success_version as u64 + 1),
```

Comment:
I have to leave in the `+1`s unfortunately since you can't factor it out from the different logical branches. I renamed the function to `get_starting_version_from_db` to correct the misnomer.